### ISSUE TITLE
Show linter output for closed files

### ISSUE THREAD

**Issue by mheinzler** [SublimeLinter/SublimeLinter#1607]
Hi guys,

I know it has been some time but I finally wanted to finish the mess I started :smile: and show the output of a multi-file linter also for closed files.

I was looking through the code to see how this could be done but I'm not really sure what the best way is. All views seem to depend on a bid, so my first thought was to use a special bid value to signal that a file is closed and requires special handling. Or something like a bid/filename pair as the key to store errors. These could then be updated later when the files are opened.

Any ideas?

**Comment by kaste** [comment]
Haha, I'm still doing the refactorings after your mess 😀 

I think this is relatively straight: We introduce a new store at `persist` for the transition. 

```
# persist.py
errors = defaultdict(list)  # type: DefaultDict[sublime.BufferId, List[LintError]]
new_error_store = defaultdict(list)  # type: DefaultDict[Filename, List[LintError]]
```

- We first ensure that this new store gets populated correctly.  
- Wherever we use the old store `persist.errors.get(bid)` we replace it with `new_error_store.get(filename)`. Since we're event driven here, and we usually have a view which we ask for `buffer_id()`. We can ask for `file_name()` as well. E.g. 

```
# goto_commands.py
def goto(view, direction, count, wrap):
    bid = view.buffer_id()
    errors = persist.errors.get(bid)
```

looks like a simple transformation.

This should work okay for the panel which is essentially per window anyway.

For the regions (`highlight_view.py`), we probably just listen for `on_load_async(self, view)` events, and thus draw/update after load.

The problematic parts:

We probably need to observe renames, esp. likely from unnamed buffers to saved ones, and update the store after that. 

We need to discuss, an idea when we actually throw away/cleanup the store. Right now, we always cleanup 'on_post_close' (after that the 'bid' is invalid anyway!), but since the filenames are more stable than bids we could keep the errors for longer periods as well. But ... every 'git checkout random_branch' could lead to orphaned files, also deleting a file, closing an unsaved buffer. 

I think we should discuss these issues first bc it changes the behavior of SublimeLinter.

**Comment by mheinzler** [comment]
This sounds good. If you like I can draft a PR and start with implementing the new error store.

In my main use case (C++) all errors from other files are basically linked to the linted file. Here it would make sense to handle these errors the same way as if they had come from the linted file. So if the linted file is updated/closed, also remove the errors that were reported for other files. I think this already is the current behavior.

For linters that process the whole project there could be a setting which specifies that all errors are independent of each other. In that case the store needs to be cleared manually.

**Comment by kaste** [comment]
Sure, go for it! Start on top of #1595 bc it already changes some type signatures in sublime_linter.py you also touch.

**Comment by Darkle** [comment]
Could this be left open so I could follow the progress?

Ta.

**Comment by kaste** [comment]
@Darkle This feature is done and has shipped with 4.13
